DeeperFunc/Router.swift

```diff
@@ -0,0 +1,191 @@
+                .flatMap({ $0.rest.path.isEmpty ? intent : nil })
+        })
+        templates.append(route.template)
+        return self
+    }
+    
+    @discardableResult
+    public func add(_ intent: U, format: String) -> Router {
+        guard let route = format.routePattern else { return self }
+        add({ url in
+            route.parse(routeComponents(from: url))
+                .flatMap({ $0.rest.path.isEmpty ? intent : nil })
+        })
+        templates.append(route.template)
+        return self
+    }
+
+    @discardableResult
+    public func add<A, S: ClosedPatternState>(_ intent: @escaping (A) -> U, _ route: RoutePattern<A, S>) -> Router {
+        add({ url in
+            route.parse(routeComponents(from: url))
+                .flatMap({ $0.rest.path.isEmpty ? $0.match : nil })
+                .map(intent)
+        })
+        templates.append(route.template)
+        return self
+    }
+
+    @discardableResult
+    public func add<A>(_ intent: @escaping (A) -> U, format: String) -> Router {
+        guard let route = format.routePattern else { return self }
+        add({ url in
+            route.parse(routeComponents(from: url))
+                .flatMap({ $0.rest.path.isEmpty ? $0.match : nil })
+                .flatMap({ $0 as? A })
+                .map(intent)
+        })
+        templates.append(route.template)
+        return self
+    }
+
+    @discardableResult
+    public func add<A, B, C, S: ClosedPatternState>(_ intent: @escaping (A, B, C) -> U, _ route: RoutePattern<((A, B), C), S>) -> Router {
+        add({ url in
+            route.parse(routeComponents(from: url))
+                .flatMap({ $0.rest.path.isEmpty ? $0.match : nil })
+                .map(flatten)
+                .map(intent)
+        })
+        templates.append(route.template)
+        return self
+    }
+    
+//    @discardableResult
+//    public func add<A, B, C>(_ intent: @escaping (A, B, C) -> U, format: String) -> Router {
+//        let route = format.routePattern
+//        add({ url in
+//            route.parse(routeComponents(from: url))
+//                .flatMap({ $0.rest.path.isEmpty ? $0.match : nil })
+//                .flatMap({ $0 as? ((A, B), C) })
+//                .map(flatten)
+//                .map(intent)
+//        })
+//        templates.append(route.template)
+//        return self
+//    }
+//
+    @discardableResult
+    public func add<A, B, C, D, S: ClosedPatternState>(_ intent: @escaping (A, B, C, D) -> U, _ route: RoutePattern<(((A, B), C), D), S>) -> Router {
+        add({ url in
+            route.parse(routeComponents(from: url))
+                .flatMap({ $0.rest.path.isEmpty ? $0.match : nil })
+                .map(flatten)
+                .map(intent)
+        })
+        templates.append(route.template)
+        return self
+    }
+
+    @discardableResult
+    public func add4<A, B, C, D>(_ intent: @escaping (A, B, C, D) -> U, format: String) -> Router {
```

Comment:
using `add` for all function variants results in ambiguity (why?), so had to add this 4 as a suffix

Reply:
Solved by adding extra pair of brackets around method variant for single parameter